### PR TITLE
Adding Extra Uart peripheral to LM4F120.

### PR DESCRIPTION
**Pull request by rakendrathapa**
Added support for an extra UART(UART_1) peripheral for Stellaris ek-lm4f120xl board.
Also, there is certain code restructuring for better understanding and clarity.

Regards,
Rakendra

### CODE REVIEW

**Comment by thomaseichinger** [comment]
@rakendrathapa Thank you for your contribution. On first sight the code looks good, great work. In #3787 @dkm mentioned rewriting the timer implementation, would it be possible you two synchronise on this? Did you run the tests in `tests` directory regarding timers? Did you experience any issues?

**Comment by dkm** [comment]
Sure ! The easiest thing would be to remove the TIMER1 part from this PR as I'll add it in my timer-fix PR. What do you think ?

**Comment by rakendrathapa** [comment]
On Mon, Oct 12, 2015 at 2:24 PM, Marc Poulhiès <notifications@github.com> wrote:

    Sure ! The easiest thing would be to remove the TIMER1 part from this PR as I'll add it in my timer-fix PR. What do you think ?

**Comment by rakendrathapa** [comment]
Hi,

@thomaseichinger  Yes I ran tests in tests directory. The results were as expected. And I will sync up with Marc on this.

@dkm  Okay. l will remove the Timer1 update from this PR. You add it in your PR.
Meanwhile, I took a look at your repository. Glad to see other periph updated :)
Let me know, when your working on ADC, GPIO and PWM are complete. I want to start working with them and will test them in parallel.

Thanks and Regards,
Rakendra

**Comment by thomaseichinger** [comment]
@dkm @rakendrathapa I think another option would be if @dkm opens a PR against @rakendrathapa's branch, this may minimise coding and review needs. Commit authors would also be persistent.

**Comment by dkm** [comment]
My changes will basically rewrite everything in lm4f120's timer.c, so I'm not sure this would help: I won't keep much (if anything...) from @rakendrathapa timer changes...

**Comment by rakendrathapa** [comment]
@dkm  Y the whole rewrite ? Also did you take a look at new timer routine in this PR, there are lot of restructuring.
Also, I think what @thomaseichinger is suggesting is a good option. You can open the PR for the timer routine against my branch and we can collaborate and work on it.

**Comment by rakendrathapa** [comment]
There is a bug with current timer routine which I found while running some tests yesterday. Will be checking in the changes soon.

**Comment by dkm** [comment]
I'm not rewriting everything, but timer logic has to change... RIOT expect to be able to change timer tick rate, which is not possible on this platform (I use software scaling...). Prescaler only extends timer range so you can change IT freq, not timer increase rate which happens at sysclock... So I've revamped timer's logic wrt to timer_init & xtimer. Also, xtimer expects underlying timer to keep its period and use matching IT for intermediate timer, which is not the case in current implem.

**Comment by rakendrathapa** [comment]
I agree. Current implementation has issues with xtimer. So, how are we planning to proceed with it. Opening PR for timer implementation against my branch or a new PR ?

**Comment by dkm** [comment]
I can open a PR on your branch, I don't mind. My branch is still WIP, but it seems to work correctly. You can expect a ping later today or tomorrow :)
Then I'll go back to ADC, PWN and then SPI...

**Comment by dkm** [comment]
Sorry, wanted to do a PR on your branch, but could not find how. github gives me the long list of RIOT fork, but your's was not listed (or not at raken*). So I've opened a new PR.

**Comment by rakendrathapa** [comment]
@dkm No issues :) We can work with your PR.

**Comment by PeterKietzmann** [comment]
So can we close this PR in favour of #4089 ?

**Comment by dkm** [comment]
Not really. #4089 only fix timers. This PR does not fix timers, but adds uart. If @rakendrathapa  could simply remove timers changes from this PR, that would remove the overlap !

**Comment by rakendrathapa** [comment]
Sorry for my sloppiness on this. I had been travelling last two weeks and couldn't update this. I will update the PR with UART only changes soon.

**Comment by rakendrathapa** [comment]
Have removed Timer changes.

**Comment by rakendrathapa** [comment]
@thomaseichinger @PeterKietzmann Can the CL flag be enabled to ensure that Travis and strider are success.

**Comment by PeterKietzmann** [comment]
I can set it, yes. But I tend to wait for @thomaseichinger's ACK before merge

**Comment by authmillenon** [comment]
@thomaseichinger ping?

File: cpu/lm4f120/periph/uart.c [hunk 1]
```diff
@@ -28,6 +28,15 @@
 #include "periph_conf.h"
 
 /**
+ * @brief Unified interrupt handler for all UART devices
+ *
+ * @param uartnum       the number of the UART that triggered the ISR
+ * @param ulBase        the UART device that triggered the ISR
+ * @param ulRxInt       the RX Interrupt Mask of the UART Device that triggered the ISR
+ */
+static inline void irq_handler(uart_t uartnum, unsigned long ulBase, unsigned long ulRxInt);
```
Comment (thomaseichinger) [inline cpu/lm4f120/periph/uart.c]: Here and all following, please don't use camel case. https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#naming

**Comment by thomaseichinger** [comment]
I'm fine with the code, @rakendrathapa please adapt coding conventions and we should be good to go.

**Comment by emmanuelsearch** [comment]
@rakendrathapa ping

**Comment by rakendrathapa** [comment]
Yea..it is :)
Have been a bit preoccupied by other work since last couple of weeks which I am afraid may continue for this week as well.  :(  
Just need to make myself sit for this. But this will definitely be ready for merging by end of this month. 

Regards,
Rakendra

**Comment by dkm** [comment]
I guess https://github.com/RIOT-OS/RIOT/pull/4393 could be merged also... I'm not rebasing the branch anymore because of lack of response, but as soon as a dev give me a sign, i'm more than motivated :)

**Comment by OlegHahm** [comment]
@rakendrathapa, willing to rebase after #4393 got merged? (Sorry for the delay.)

**Comment by rakendrathapa** [comment]
Sorry for the delay on my part for this. 
@OlegHahm Yes, the branch has now been updated with the recent changes.

Regards

**Comment by emmanuelsearch** [comment]
Willing to test. How should I proceed?

**Comment by haukepetersen** [comment]
needs rebase

**Comment by emmanuelsearch** [comment]
Tested this with @haukepetersen 
It seems that TX works, but RX breaks.
Can you confirm that it works both ways for you?

**Comment by rakendrathapa** [comment]
@haukepetersen  Rebased..
@emmanuelsearch Thanks for your patience on this. I will debug it.

Regards

**Comment by rakendrathapa** [comment]
I see nhdp test case failing which is failing travis cl build. When I locally try to see the error, this throws error: 
"unable to access 'http://olsr.org/git/oonf.git/': The requested URL returned error: 403"
Can anyone guide me how to make it work ? 

Regards

**Comment by cgundogan** [comment]
there's an open issue for that: #5107

**Comment by rakendrathapa** [comment]
@emmanuelsearch @haukepetersen 
I tried debugging it, but I see both TX and RX working.

For testing purpose, I used an earlier test case "perip_uart_int" and modified it for current testing scenario as it is simple and straightforward.

Test case: The test reads characters from the UART into a receiving buffer. The received string is printed to stdout. In the same time, a string is written to the UART every 2 seconds, to make sure the transmission

Result: As I connect U1TX (PB1) and U1RX (PB0), the transmitted string is received and outputted to stdout.

In case, required for testing, the test case is present in my local repository branch: periph_uart_test. 
Directory: RIOT/examples/periph_uart_test/

Do let me know the test scenario where Rx failed, will use it to debug. 

Thanks and Regards,
Rakendra

**Comment by rakendrathapa** [comment]
@cgundogan Oh okay. Thanks :+1:

**Comment by cgundogan** [comment]
@rakendrathapa could you please rebase your branch to the latest master? This should fix the NHDP/oonf_api issues.

**Comment by rakendrathapa** [comment]
@emmanuelsearch @haukepetersen 
Any thoughts ??

**Comment by rakendrathapa** [comment]
@haukepetersen @emmanuelsearch .. Any feedback on the same. I tested with test/periph_uart and I see TX and  RX working fine. Can you please take a look ?

**Comment by kYc0o** [comment]
Since this is a new feature, and we are on feature freeze, this PR will be postponed. Please rebase and let's try to merge it for the next release.

**Comment by dkm** [comment]
Ok, I'll do this... later...

File: cpu/lm4f120/periph/uart.c (hunk 1 above)
Comment (cgundogan) [inline cpu/lm4f120/periph/uart.c]: @rakendrathapa could you remove the camel case here like pointed out by @thomaseichinger (:

**Comment by miri64** [comment]
Postponed due to Feature Freeze

**Comment by PeterKietzmann** [comment]
@rakendrathapa can you address the comment concerning CamelCase and do a rebase? @emmanuelsearch any chance giving this a try again?

**Comment by kYc0o** [comment]
This PR needs rebase and maybe a final test, but I don't have the board at the moment. So I propose to, unfortunately, postpone it again...

**Comment by dkm** [comment]
Getting something merged for a low traffic board like this one is really hard. @rakendrathapa did not touch the code since last April.
Still not possible to have a set of «unmaintained» boards ? Where riot team don't need to test, but "only" do code review. ?

**Comment by kYc0o** [comment]
I reviewed the code and besides the camel case issue, which is almost everywhere, the code on cpu/lm4f120/periph/uart.c seems very outdated.

So unless @rakendrathapa updates this PR it cannot be merged.

For the testing part we sometimes merge code which looks OK even if we don't have the board, and for this case actually we have it, just I don't have it right now. We try to test PRs as much as possible, but I don't think "not merge because we don't have the board" is a policy.

**Comment by dkm** [comment]
Thank you for taking the time to reply :)
I guess I had a bad exp when trying to add SPI on this board, as it has been left aside because nobody except me could test it (https://github.com/RIOT-OS/RIOT/pull/5412 ). If that's not a policy, then maybe we'll be able to make this PR up-to-date and merged. I have to refresh the SPI part, maybe I could take over this one if @rakendrathapa is not available.

**Comment by rakendrathapa** [comment]
Sorry for all the delays. I am not working on this board anymore. 
But I will try to get this small change done by tonight so that it does not block others.

**Comment by rakendrathapa** [comment]
@dkm .. If you can take it up that will be great..it has been quite some time that I have worked on the board.

**Comment by kYc0o** [comment]
> If that's not a policy, then maybe we'll be able to make this PR up-to-date and merged. I have to refresh the SPI part

Sounds great!

**Comment by dkm** [comment]
@rakendrathapa not sure exactly what are your plan: 1) will you try to refresh the patch tonight, or 2) do you let me take over the job ? I would prefer 1), but can still do 2) (will take longer, but time is not really an issue, it seems I'm the only one still using the board)

**Comment by rakendrathapa** [comment]
@dkm If you can take it that will be better..
I have not worked on the board since quite some time..patch testing may take some time.

**Comment by dkm** [comment]
@rakendrathapa ok. Will refresh my setup tomorrow and try to get everything cleaned.

**Comment by dkm** [comment]
?

**Comment by rakendrathapa** [comment]
@dkm Oh you can take it over. I closed it since I wasn't planning to start work on it anytime soon.
Do you want me to reopen the PR.. ?

**Comment by dkm** [comment]
Let's leave it like this. If I want to work on it, I'll reopen it.

**Comment by rakendrathapa** [comment]
Cool ..

**Pull request closed without merging**